[DRAFT] Initial version of the boundary event documentation (#1)
DmytroKost/docs: content/en/docs/refguide/modeling/application-logic/workflows/workflow-elements/boundary-events/timer-event.md

---
title: "Timer Event"
url: /refguide/timer-event/
weight: 10
---

{{% alert color="info" %}}
This feature is currently in public beta as of version 10.15. Please note that certain features or attributes may be subject to change and may contain bugs.{{% /alert %}}

## Introduction

The **Timer Event** enables the timed-execution of alternative flows when the guarded activity is exceeding the set timer.

## Properties

**Timer Event** properties consist of the following sections:

* [General](#general)
* [Timer](#timer)
* [Common](#common)

### General Section {#general}

> The **Caption** describes what happens in this element. It is displayed in the workflow element to make the **Timer** event easier to read and understand without the need to add annotations.

Reviewer: ```suggestion
The **Caption** describes what happens in this element. It is displayed on the workflow element to make the **Timer** event easier to read and understand.
```

### Timer Section {#timer}

> The **Timer** property defines at which time the workflow path continues. With this property, you can set the timer in two ways: you can set a certain duration, or a certain date and time with an expression. When an activity before the timer finishes and the workflow path reaches the timer, the configured duration or date and time will be scheduled to take effect.

Reviewer: ```suggestion
The **Timer** property specifies the exact moment or time interval after which the workflow progresses to the next step. With this property, you can set the timer in two ways: 
1. you can set a certain duration
2. certain date and time with an expression

When an activity finishes before the **Timer** has elapsed, and the workflow path reaches the timer, the configured duration or date and time will be scheduled to take effect.
```

I don't really understand the last sentence ☝️ 

Also, for the numbered items, I think we usually describe the Studio Pro options verbatim, followed by a description. Perhaps Yiyun has some suggestions here.

| Expression | You can set a certain date and time for the timer by writing an expression via the **Continue at** setting.<br><br>For example, you can write `addDays([%CurrentDateTime%], 1)` to set tomorrow as the due date and time. To set a static date and time, you can use the expression `parseDateTimeUTC('2023-12-10T17:12:00.000', 'yyyy-MM-dd''T''HH:mm:ss.SSS')`.<br><br>You can also create a more complex timer. For example, you can set a timer based on a Boolean value (in this example, `isVIPUser`) from the provided workflow context entity: `if $WorkflowContext/isVIPUser then addDays([%CurrentDateTime%], 2) else addWeeks([%CurrentDateTime%], 2])`.<br><br>For more information on available expressions in Mendix, see [Expressions](/refguide/expressions/). |

## Timer Expiration {#timer-expiration}

When a **Timer** event expires, it behaves differently depending on the state of the workflow:

> * When a timer event is set on an in-progress workflow, the workflow continues when the timer expires, and continuing the modeled flow.

Reviewer: ```suggestion
* When a timer event is set on an in-progress workflow, the workflow continues when the timer expires, and will continue the flow.
```

I think this point is a bit misleading as it now reads as if you can set the timer on a workflow itself.

> * When a timer event is set on a paused workflow and when the timer expires, the workflow only starts the modeled flow after the workflow is set to in progress again.

Reviewer: I don't think we need to use the word "modeled"

### Workflow Incompatibility

When a **Timer** event is added to the workflow definition and the application is redeployed, a validation on already running workflow instances is performed. When the **Timer** event has been added before the currently in-progress activity, the workflow becomes incompatible. The conflict/incompatibility validation is analogous to other activities added before an in-progress activity. For more information, see [Workflow Versioning and Conflict Mitigation](/refguide/workflow-versioning/).

When a **Timer** event is removed from the workflow definition and the application is redeployed, on initiation of the application, it validates if there are any running timers (that is, active timers that are initiated but have not reached their defined date and time). In this case, the workflow becomes incompatible and a warning log is created. For information on how to resolve a conflict when an activity is removed, see [Workflow Versioning and Conflict Mitigation](/refguide/workflow-versioning/).

### Specific Workflow State Cases

The following cases do not trigger a start of the boundary event path when timer expires.

> * Expiration in a workflow that is aborted.

Reviewer: I think we can leave out `Expiration...` as that is implied by the preceding sentence.

---
title: "Timer Event"
url: /refguide/timer-event/
weight: 10
---

{{% alert color="info" %}}
This feature is currently in public beta as of version 10.15. Please note that certain features or attributes may be subject to change and may contain bugs.{{% /alert %}}

## Introduction

> The **Timer Event** enables the timed-execution of alternative flows when the guarded activity is exceeding the set timer.

Reviewer 2: replace "guarded" with parent (everywhere).

Author: changed it everywhere

## Properties

**Timer Event** properties consist of the following sections:

* [General](#general)
* [Timer](#timer)
* [Common](#common)

### General Section {#general}

> The **Caption** describes what happens in this element. It is displayed on the workflow element to make the **Timer** event easier to read and understand.

Reviewer 3: If we don't treat it as a workflow "element" we probably should mention it being such. So, we should reword it into something like "describes what happens in this boundary event"

| Expression | You can set a certain date and time for the timer by writing an expression via the **Continue at** setting.<br><br>For example, you can write `addDays([%CurrentDateTime%], 1)` to set tomorrow as the due date and time. To set a static date and time, you can use the expression `parseDateTimeUTC('2023-12-10T17:12:00.000', 'yyyy-MM-dd''T''HH:mm:ss.SSS')`.<br><br>You can also create a more complex timer. For example, you can set a timer based on a Boolean value (in this example, `isVIPUser`) from the provided workflow context entity: `if $WorkflowContext/isVIPUser then addDays([%CurrentDateTime%], 2) else addWeeks([%CurrentDateTime%], 2])`.<br><br>For more information on available expressions in Mendix, see [Expressions](/refguide/expressions/). |

## Timer Expiration {#timer-expiration}

When a **Timer** event expires, it behaves differently depending on the state of the workflow:

* * When a timer event is set on an in-progress workflow, the workflow continues when the timer expires, and will continue the flow.

> * When a timer event is set on a paused workflow and when the timer expires, the workflow only starts the flow after the workflow is set to in progress again.

Reviewer 3: This item refers to the behavior from `Specific Workflow State Cases`, where it is described in more detail. I think we should combine the sections, or move the items from  `Specific Workflow State Cases` to the `Timer Expiration` section

> ### Workflow Incompatibility

Reviewer 3: I think this section should be removed, since addition or removal of timer events won't lead to workflow becoming incompatible if the timer hasn't expired. The section is relevant to the Time activity and the behavior that's described there.

Reviewer 3: We could mention that after the activity has started and you add more timer events they won't be scheduled. But that's not about workflow incompatibility. It probably belongs to the `Specific Workflow State Cases` section

Author: Removed this section
